File: Graphs/TopologicalSort.java
// File Name: TopologicalSort.java

import java.util.*;

public class TopologicalSort {

    static ArrayList<ArrayList<Integer>> graph = new ArrayList<>();
    static boolean[] visited;
    static Stack<Integer> stack = new Stack<>();

    static void dfs(int v) {
        visited[v] = true;

        for (int neighbor : graph.get(v)) {
            if (!visited[neighbor]) {
                dfs(neighbor);
            }
        }

        stack.push(v);
    }

    static void topologicalSort(int vertices) {
        for (int i = 0; i < vertices; i++) {
            if (!visited[i])
                dfs(i);
        }

        System.out.println("Topological Order:");
        while (!stack.isEmpty())
            System.out.print(stack.pop() + " ");
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter number of vertices: ");
        int vertices = sc.nextInt();

        System.out.print("Enter number of edges: ");
        int edges = sc.nextInt();

        visited = new boolean[vertices];

        for (int i = 0; i < vertices; i++)
            graph.add(new ArrayList<>());

        System.out.println("Enter directed edges (u v):");
        for (int i = 0; i < edges; i++) {
            int u = sc.nextInt();
            int v = sc.nextInt();
            graph.get(u).add(v);  // Directed edge
        }

        topologicalSort(vertices);
    }
}

File: Graphs/topological_sort.py
# File Name: topological_sort.py

def dfs(v, visited, stack, graph):
    visited[v] = True

    for neighbor in graph[v]:
        if not visited[neighbor]:
            dfs(neighbor, visited, stack, graph)

    stack.append(v)

vertices = int(input("Enter number of vertices: "))
edges = int(input("Enter number of edges: "))

graph = [[] for _ in range(vertices)]
visited = [False] * vertices
stack = []

print("Enter directed edges (u v):")
for _ in range(edges):
    u, v = map(int, input().split())
    graph[u].append(v)

for i in range(vertices):
    if not visited[i]:
        dfs(i, visited, stack, graph)

print("Topological Order:")
while stack:
    print(stack.pop(), end=" ")
